src/stores/redis.rs

```diff
@@ -0,0 +1,436 @@
+            Some(val) => {
+                self.hits += 1;
+                self.base_vec_fill(&val)
+            }
+            None => {
+                self.misses += 1;
+                let val = f();
+                self.base_set(&key, &val);
+                let index = self.store.len();
+                self.store.push(val);
+                &mut self.store[index]
+            }
+        }
+    }
+
+    fn cache_remove(&mut self, key: &K) -> Option<V> {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        self.base_get(&key).map(|val| {
+            con.del::<String, ()>(self.generate_key(key)).unwrap();
+            serde_json::from_str(&val).unwrap()
+        })
+    }
+
+    fn cache_clear(&mut self) {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        redis_rs::cmd("EVAL")
+            .arg("local keys = redis.call('keys', ARGV[1]) \n for i=1,#keys,5000 do \n redis.call('del', unpack(keys, i, math.min(i+4999, #keys))) \n end \n return keys")
+            .arg(0)
+            .arg(format!("{}*", self.generate_prefix()))
+            .query(&mut con)
+            .unwrap()
+    }
+
+    /// In `RedisCache`, it's an alias to `cache_clear`
+    fn cache_reset(&mut self) {
+        self.cache_clear();
+    }
+
+    fn cache_size(&self) -> usize {
+        let mut con = self.client.get_connection().unwrap();
+        redis_rs::cmd("EVAL")
+            .arg(format!(
+                "return #redis.call('keys', '{}*')",
+                self.generate_prefix()
+            ))
+            .arg(0)
+            .query(&mut con)
+            .unwrap()
+    }
+
+    fn cache_hits(&self) -> Option<u64> {
+        Some(self.hits)
+    }
+
+    fn cache_misses(&self) -> Option<u64> {
+        Some(self.misses)
+    }
+
+    fn cache_lifespan(&self) -> Option<u64> {
+        self.seconds
+    }
+
+    fn cache_set_lifespan(&mut self, seconds: u64) -> Option<u64> {
+        self.clean_up();
+        let old = self.seconds;
+        self.seconds = Some(seconds);
+        old
+    }
+}
+
+#[cfg(feature = "async")]
+#[async_trait]
+impl<K, V> CachedAsync<K, V> for RedisCache<K, V>
+where
+    K: Send + Display,
+    V: Serialize + DeserializeOwned + Clone,
+{
+    async fn get_or_set_with<F, Fut>(&mut self, key: K, f: F) -> &mut V
+    where
+        V: Send,
+        F: FnOnce() -> Fut + Send,
+        Fut: Future<Output = V> + Send,
+    {
+        let mut con = self.client.get_connection().unwrap();
+        let generated_key = self.generate_key(&key);
+        let val: RedisResult<String> = con.get(generated_key.clone());
+        match val {
+            Ok(val) => {
+                self.hits += 1;
+                self.base_vec_fill(&val)
+            }
+            Err(_) => {
+                let val = f().await;
+                self.base_set(&key, &val);
+                let index = self.store.len();
+                self.store.push(val);
```

Comment:
what's the reason for also storing the value in a vec?

Reply:
To return a reference to it, since it's in the trait signature. If you know a better way, please don't hesitate.